### PR TITLE
Add torch.nn.functional.threshold ref

### PR DESCRIPTION
**Pull request by wconstab**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -20322,6 +20326,16 @@ def __init__(
             DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_executor'),
         ),
     ),
+    ElementwiseUnaryPythonRefInfo(
+        "_refs.nn.functional.threshold",
+        torch_opinfo_name="nn.functional.threshold",
+        skips=(
+            # NotImplementedError: Could not run 'aten::where.ScalarOther' with arguments from the 'AutogradMeta'...
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_meta',),
```
Comment (wconstab) [inline torch/testing/_internal/common_methods_invocations.py:20334]: @mruberry just out of curiosity, what's the difference between marking expected test failures here vs in the UnaryUfuncInfo above?  (One difference is that only here seemed to work)

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -20322,6 +20326,16 @@ def __init__(
...
+            # RuntimeError: Tracing expected 3 arguments but got 1 concrete arguments
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_executor'),
```
Comment (wconstab) [inline torch/testing/_internal/common_methods_invocations.py:20336]: @mruberry is it expected that I need to xfail this test? 

For the ref_meta one i think there is a missing meta kernel, but for this one i'm not sure what the reason is.

File: torch/_refs/nn/functional/__init__.py [hunk 3]
```diff
@@ -332,6 +333,27 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
...
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.threshold
+    """
+
+    if inplace:
+        raise NotImplementedError
+
+    # the straightforward computation drops nan's from input in some cases
+    partial = torch.where(a > threshold, a, value)
```
Comment (wconstab) [inline torch/_refs/nn/functional/__init__.py:352]: @mruberry i think threshold does not support complex numbers (based on trial with existing eager), but there doesn't seem to be an analogue of 'NumberType' that doesn't include complex.  Should I just assert value is not complex and hope that is enough to make mypy happy, or is there any other pattern i can follow here?

Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:352]: You can do it in one step
```
torch.where(a <= threshold, value, a)
```
which will propagate all nan values from `a`. I'm not certain if it should be `<=` or `<` here, check the grad behavior.

Comment (wconstab) [reply torch/_refs/nn/functional/__init__.py:352]: oh, is nan propagation asymetric somehow? i did try to do it in one step but with the `>` comparison direction, and nans definitely didn't propagate.

Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py]: Comparison ops when one of the args is `nan` always return `false` so (given that you want only propagate `nans` from `a`) `a` should be on the lhs side of `where`.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:20334]: These skips apply to `_refs.nn.functional.threshol`, while the other skips would apply to `nn.functional.threshold`

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:20336]: This should be fixed after https://github.com/pytorch/pytorch/pull/79858 from @ngimel, but yes its' fine to xfail this for now

File: torch/_refs/nn/functional/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/nn/functional/__init__.py]: You can do that, and/or you an explicitly annotate the type with `Union[int, float]` or `Union[bool, int, float]` as appropriate, too

**Comment by mruberry** [comment]
Looking good overall Will -- let me know if you have any additional questions!

**Comment by ngimel** [comment]
Please wait with merging until #79858 lands (or skip the test instead of xfailing)

**Comment by wconstab** [comment]
oops, i missed that conversation, just rebased on master including #79858 and re-enabled the test now.  don't have a local build ready so i am just waiting to see if CI is happy.

File: torch/_refs/nn/functional/__init__.py [hunk 4]
```diff
@@ -357,6 +358,27 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
...
+    inplace: bool = False,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.threshold
+    """
+
+    if inplace:
+        raise NotImplementedError
+
+    return torch.where(a <= threshold, value, a)
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:379]: Tie goes to the value

**Comment by mruberry** [comment]
Failures are tests like test_reference_numerics_extremal_nn_functional_threshold_cpu_float32 which should reproduce locally. It seems like the NaN propagation is divergent from eager PyTorch.

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -14818,18 +14818,22 @@ def error_inputs_mean(op_info, device, **kwargs):
                          active_if=(IS_MACOS or IS_WINDOWS)),
         ),
     ),
-    OpInfo(
+    UnaryUfuncInfo(
         'nn.functional.threshold',
-        aten_backward_name='threshold_backward',
         ref=lambda x, threshold, value: np.where(x > threshold, x, value).astype(x.dtype),
```
Comment (wconstab) [inline torch/testing/_internal/common_methods_invocations.py:14823]: @mruberry @ngimel  I'm not sure what to aim for here.

1) my `UnaryUfuncInfo` preserves the existing `ref= ... np.where(x > threshold, ...` 
2) my _refs.nn.functional.threshold op seems to match the numpy ref
3) torch.nn.functional.threshold doesn't match either the numpy ref or my ref for nan handling

So.. should I be treating torch.nn.functional.threshold as the gold standard and perhaps update the numpy ref to match it? (If so i am not sure why the orig Opinfo test didn't catch this: oh- i guess by making a UnaryUfuncInfo I am opting into new test coverage for the first time?)

This makes a little more sense after i wrote it out, but i'm still not sure what the best change would be.

Comment (ngimel) [reply torch/testing/_internal/common_methods_invocations.py:14823]: I'm confused, to me it looks like torch.nn.functional.threshold and Will's decomposition produce the same result (numpy ref might be different):
```
In [2]: inf=float('inf')

In [3]: nan=float('nan')

In [4]: a=torch.tensor([inf, inf, inf, -inf, -inf, -inf, nan, nan, nan])

In [5]: threshold=-0.123

In [6]: value=9.

In [7]: torch.nn.functional.threshold(a, threshold, value)
Out[7]: tensor([inf, inf, inf, 9., 9., 9., nan, nan, nan])

In [9]: torch.where(a <= threshold, value, a)
Out[9]: tensor([inf, inf, inf, 9., 9., 9., nan, nan, nan])
```

Comment (wconstab) [reply torch/testing/_internal/common_methods_invocations.py:14823]: > I'm confused, to me it looks like torch.nn.functional.threshold and Will's decomposition produce the same result (numpy ref might be different):

yes, i agree with this @ngimel 

>my _refs.nn.functional.threshold op seems to match the numpy ref
torch.nn.functional.threshold doesn't match either the numpy ref or my ref for nan handling

in my statement above, i had modified my _refs version to match the numpy ref, hence the different conclusion.

It sounds like the numpy ref is the one that needs to change, so i'll try to get that to match nn.functional first and then see how my _refs version looks.

**Comment by wconstab** [comment]
ok fingers crossed for a clean CI this time. should i opt into any other jobs or is the default PR coverage going to hit all the things that use the numpy ref and _refs op?

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -14758,18 +14758,22 @@ def error_inputs_mean(op_info, device, **kwargs):
                          active_if=(IS_MACOS or IS_WINDOWS)),
         ),
     ),
-    OpInfo(
+    UnaryUfuncInfo(
         'nn.functional.threshold',
-        aten_backward_name='threshold_backward',
-        ref=lambda x, threshold, value: np.where(x > threshold, x, value).astype(x.dtype),
+        ref=lambda x, threshold, value: np.where(np.isnan(x), x, np.where(x > threshold, x, value)).astype(x.dtype),
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:14763]: why don't you use the same numpy reference as your ref decomposition? 
```suggestion
        ref=lambda x, threshold, value: np.where(x <= threshold, value, x).astype(x.dtype),
```

Comment (wconstab) [reply torch/testing/_internal/common_methods_invocations.py]: yea, good call. i thought i had tried that without success but it indeed works.

**Comment by mruberry** [comment]
This looks good and the test failure seem irrelevant but it's picked up some ONNX changes that need to be removed

**Review by mruberry — APPROVED** [review]
Cool! But be sure to remove the third_party/onnx submodule update from this PR

**Pull request closed without merging**